feat(Chat): add different chat item types

mnajdova opened the pull request:
# ChatMessage

This PR aims to add different types of messages for the chat. We should support the following types:
- control messages 
![image](https://user-images.githubusercontent.com/4512430/45810105-267de280-bcca-11e8-891c-ffc6ef168085.png)
- bubble messages (this is currently the implementation of the ChatMessage itself)
![image](https://user-images.githubusercontent.com/4512430/45810165-4f9e7300-bcca-11e8-9dcb-6b99684133c6.png)
- divider messages
![image](https://user-images.githubusercontent.com/4512430/45810196-5fb65280-bcca-11e8-8c09-4aadb68875c4.png)


### TODO

- [x] Conformance test
- [ ] Minimal doc site example
- [ ] Stardust base theme
- [x] Teams Light theme
- [ ] Teams Dark theme
- [ ] Teams Contrast theme
- [ ] Confirm RTL usage
- [ ] [W3 accessibility](https://www.w3.org/standards/webdesign/accessibility) check
- [ ] [Stardust accessibility](https://github.com/stardust-ui/accessibility) check
- [ ] Update glossary props table
- [x] Update the CHANGELOG.md

I have three proposals and would like to hear your opinion on.

# API Proposal 1

## type: 'bubble' | 'control' | 'divider'

The type property will be used for rendering one of the three different components. The properties will remain the same, we would just add additional icon shorthand for the control component.

## Usage:
```jsx
<ChatMessage type='bubble' author={...} ... />
<ChatMessage type='control' icon={...} ... />
<ChatMessage type='divider' content={'Yesterday'} ... />
```
This would be easy for use, but not much open for fully customizing the generated component.

# API Proposal 2

## bubble: shorthand for the Message.Bubble component

## control: shorthand for the Message.Control component

## divider: shorthand for the Message.Divider component

Open for customization and consistent with the other usages of the shorthand. One problem is that, a message cannot be at the same time two of these types, so we will kind a have to decide behind the scene which would take precedence. 

## Usage
```jsx
<ChatMessage bubble={{author: 'John Doe', ...}} ... />
<ChatMessage control={{icon: {{...}}, content='Some content...'}} ... />
<ChatMessage divider={{content: '.Yesterday}}... />
```
# API Proposal 3

This proposal is combination of the previous two. We should use the type for deciding which component will be rendered, but will use the shorthand for generating the component itself.

## type: 'bubble' | 'control' | 'divider'

## bubble: shorthand for the Message.Bubble component

## control: shorthand for the Message.Control component

## divider: shorthand for the Message.Divider component

```jsx
<ChatMessage type='bubble' bubble={{author: 'John Doe', ...}} ... />
<ChatMessage type='control'  control={{icon: {{...}}, content='Some content...'}} ... />
<ChatMessage type='divider' divider={{content: '.Yesterday}}... />
```

Commit:
-adjusting the header according to the styling guides

Commit:
-added chat action, bubble components
-renamed chat message to chat item component

Commit:
-fixing linting errors

Commit:
-fixed example

Commit:
-fixed resizing of the bubble message

Commit:
-added TODO for the wrapping

bmdalex commented on docs/src/examples/components/Chat/index.tsx:
```diff
@@ -1,11 +1,9 @@
 import Types from './Types'
-import Variations from './Variations'
 
 const MessageExamples = () => (
```
I know it's out of scope of PR, but we might want to rename to `ChatExamples` to avoid confusion

mnajdova replied:
Sure, I will restructure to example containing all different chat items. Will consider this as well.

bmdalex commented on src/components/Chat/ChatItem.tsx:
```diff
@@ -0,0 +1,130 @@
+  static propTypes = {
+    as: customPropTypes.as,
+
+    /** Shorthand for the bubble message. */
```
Since `bubble`, `divider` and `action` props are mutually exclusive, we need a better way to prevent users try more than one at a time; I'd suggest something like this:
```typescript
/** Shorthand for the bubble message. */
bubble: customPropTypes.disallow(['action', 'divider']),

/** Shorthand for the divider message. */
divider: customPropTypes.disallow(['action', 'bubble']),

/** Shorthand for the control message. */
action: customPropTypes.disallow(['bubble', 'divider']),
```

mnajdova replied:
Just the thing I was missing. Thanks a lot for the suggestion Alex!

bmdalex commented on src/components/Chat/ChatItem.tsx:
```diff
@@ -0,0 +1,130 @@
+  ): React.ReactNode => {
+    const { bubble, action, divider } = this.props
+
+    return (
```
right now we assume priorities `bubble` > `action` > `divider` and some users would be confused if they try to mix them (use `bubble` and `action` at the same time and only `bubble` renders); we need to make it obvious for user these are mutually exclusive, see my suggestion above

mnajdova replied:
Taken into consideration.

bmdalex commented on src/components/Chat/ChatBubble.tsx:
```diff
@@ -93,7 +93,7 @@ class ChatMessage extends UIComponent<Extendable<IChatMessageProps>, any> {
     styles,
     variables,
-  }: IRenderResultConfig<IChatMessageProps>) {
+  }: IRenderResultConfig<IChatBubbleProps>) {
```
hmm.. if you take a look at the proposal in our internal ticket, the idea was that the `ChatBubble` would contain the body of a message + media and avatar would be part of `ChatItem`; is that achievable?

![image](https://user-images.githubusercontent.com/5442794/45876204-48469a80-bd9a-11e8-8b6a-fe062fa8b253.png)

UPDATE:
Taking a closer look, we add `avatar` prop for `ChatBubble` and `icon` prop for `ChatAction`; so if we're certain that's always the case (bubbles can have avatar and actions can have icon) I think this is a good separation and you can disregard the comment

mnajdova replied:
Yes, we needed actually different components for the structure of the different types of chat items, that's why I separate those.

bmdalex replied:
got it, makes sense, thanks!

bmdalex commented on src/components/Chat/ChatAction.tsx:
```diff
@@ -0,0 +1,133 @@
+  }: IRenderResultConfig<IChatActionProps>) {
+    const { as, icon, children } = this.props
+
+    return childrenExist(children) ? (
```
@mnajdova - it looks like the `Chat.Action` subcomponent is more complex (per @codepretty 's comments in our chats, we can have an array of actions with more complicated logic on what's rendered); I'm not even sure it falls under Stardust principles, maybe we should be more generic with the `action` prop in `Chat` component

mnajdova replied:
In my opinion, we shouldn't alter this component at this moment. We can reason for this in the following manner:
- Chat.Action can have children which can define whatever the user wants inside - in this case we won't change anything right now
- We can add actions as a list and make the Chat.Action use the Accordion component, but in that case it would be in my opinion very much Team's specific rather then common web (as @Bugaa92 said, it doesn't really fall under stardust principles)

My final thought on this is, leaving the Chat.Action component as is for this now, and address this later if we think would be necessary. First, I would like after this is merged, to try and address this in the chat pane and we will see how much complication it would bring.

Commit:
-added restrictions on the bubble, divider and action props
-fixed style for the bubble when the content contains long words

bmdalex commented on docs/src/examples/components/Chat/Types/ChatExample.shorthand.tsx:
```diff
@@ -2,25 +2,63 @@ import React from 'react'
+  },
+  {
+    key: 7,
+    action: { icon: 'record', content: 'Meeting started', timestamp: 'Today, 11:15 PM' },
```
nit:

![image](https://user-images.githubusercontent.com/5442794/45882509-a8473c00-bdae-11e8-9194-ff64412dfc3d.png)

mnajdova replied:
Changed, thanks for the suggestion!

Commit:
-added examples for the different types of chat items

Commit:
-small fixes in the description of the props

Commit:
-changed the fontSize inside the chat action component to match the Text's small size

Commit:
-updated CHANGELOG

Commit:
-adjust the styles for the icon inside the ChatAction component to match the spacing of the Avatar by default

Commit:
Merge branch 'master' into feat/chat-message-types

# Conflicts:
#	CHANGELOG.md

codepretty commented on src/components/Chat/ChatBubble.tsx:
```diff
@@ -6,20 +6,20 @@ import {
 import Text from '../Text'
 
-export interface IChatMessageProps {
+export interface IChatBubbleProps {
```
Not sure how much of this belongs in Stardust, but seems like we might be missing a couple props for message bubbles.

1. How should we handle a succession of messages? It seems to be common UI practice to show those messages different from message with dividers or other messages between them...
**From FB messenger:**
![image](https://user-images.githubusercontent.com/828692/45977744-4d585380-bfff-11e8-8b00-bc9968918c65.png)

**From Teams:**
![image](https://user-images.githubusercontent.com/828692/45977941-d1124000-bfff-11e8-8b49-7b58ff5e54bc.png)

2. What about _Edited_ or _translated_ messages?
**From Telegram:**
![image](https://user-images.githubusercontent.com/828692/45977555-d4f19280-bffe-11e8-91b1-f2bd6fd8112c.png)

**From Teams:**
![image](https://user-images.githubusercontent.com/828692/45977414-614f8580-bffe-11e8-9dbd-b12de260aaca.png)

3. Message actions and important/mentioned indicators.
**From FB messenger:**
![image](https://user-images.githubusercontent.com/828692/45977846-901a2b80-bfff-11e8-8b69-04cb08b1bfb0.png)

**From Teams:**
![image](https://user-images.githubusercontent.com/828692/45977957-da031180-bfff-11e8-9aa2-a280346eeb06.png)

levithomason replied:
1. Yes, nice call.  We currently have a `mine` prop for styling messages that are yours vs others'.  Perhaps there is a `grouped` or `consecutive` prop that we have for delineating messages that should be styled like a group or consecutively sent?

1. The edited identifier is common, however, the placement is not always common.  I'm not sure the best way to implement this due to that fact, since, right now we don't allow you to change the component "template" (where slots are rendered).  We only allow changing the contents of the slots themselves.  Here's an example of the `edited` being rendered elsewhere:

    ![image](https://user-images.githubusercontent.com/5067638/45979845-37e62800-c005-11e8-9aaf-6966c49cb907.png)
Because of this, I think we should leave it to the consumer to place this where needed.

1. As with the `edited` identifier, the consumer will place their own actions where necessary for now.  There is again more variance than consistency here that I'm not sure we can reasonably create an abstraction for.  Most likely, the consumer will want very specific implementations of these parts of a message (animations, analytics, etc).

Commit:
-removed ChatAction component
-renamed ChatBubble to ChatMessage
-added mainStyle and avatarStyle props in the ChatMessage component for styling the Layout parts
-provided Children API example for the usage of the Chat component

Commit:
Merge branch 'master' into feat/chat-message-types

# Conflicts:
#	CHANGELOG.md
#	docs/src/examples/components/Chat/Variations/ChatExampleAvatar.shorthand.tsx

Commit:
-updated CHANGELOG

Commit:
-fixed chat example

alinais commented on src/components/Chat/ChatItem.tsx:
```diff
@@ -0,0 +1,72 @@
+    /** Child content. */
+    children: PropTypes.node,
+
+    /** Additional classes. */
```
Please change the description to align to the rest of the components propTypes:
"Additional CSS class name(s) to apply."

alinais commented on src/components/Chat/ChatMessage.tsx:
```diff
@@ -6,15 +6,16 @@ import {
 import {
-  ComponentVariablesInput,
   ComponentPartStyle,
+  ComponentVariablesInput,
```
I like the ordered dependencies :)

Commit:
-changed description in props

Commit:
-small fix

Commit:
-removed avatar and contentStyle from the ChatMessage component

Commit:
.

Commit:
-reverting changes from merge

Commit:
-reverting changes from merge

Commit:
-reverted changes to Button.tsx

Commit:
-reverted changes to Button.tsx

Commit:
-reverted changes to Button.tsx

Commit:
-reverted auto sorted imports

Commit:
-fixed changelog
-removed handled props from the Chat components

Commit:
-fixing merging conflicts

Commit:
-fixes after merging

Commit:
-fixes after merging

levithomason committed:
docs: cleanup chat example

kuzhelov commented on src/components/Chat/ChatItem.tsx:
```diff
@@ -0,0 +1,70 @@
+    ElementType,
+    classes,
+    rest,
+    styles,
```
nit: seems that these two guys could be omitted, as those are not used anywhere

kuzhelov commented on src/components/Chat/ChatItem.tsx:
```diff
@@ -0,0 +1,70 @@
+    const { children, content } = this.props
+
+    return (
+      <ElementType {...rest} className={cx(classes.root, classes.content)}>
```
we have potentially unreliable styling logic here - could you, please, suggest what are our plans to address it? Would rather expect this merge being performed in styles file, and exposed as `item` styles then

mnajdova replied:
This is a wrong copy/paste error, thanks for noticing. Actually we will have just the classes.root here.

kuzhelov commented on src/components/Chat/ChatItem.tsx:
```diff
@@ -0,0 +1,70 @@
+  static displayName = 'ChatItem'
+
+  static propTypes = {
+    as: customPropTypes.as,
```
nit: description for this prop is missing (as well s for a few other components of this PR, like `ChatMessage`) - is it intentional?

mnajdova replied:
It's not intentional, will provide description for all prop types in the components added/changed in this PR.

kuzhelov commented on docs/src/examples/components/Chat/Types/ChatExample.tsx:
```diff
@@ -0,0 +1,52 @@
+import React from 'react'
```
shorthand API currently is not working correctly in docs examples - the problem is apparent if one would try to click on 'Shorthand API' button:

![image](https://user-images.githubusercontent.com/9024564/46285765-4e502e80-c585-11e8-8fe8-2a09124d1dd5.png)

Do we really want to dismiss shorthand case for `Chat`? This seems to be confusing, as by looking on `Chat`'s implementation it is clearly visible that `items` prop is exposed - thus, it is possible to utilize shorthand advantages for this component. More than that, it seems that we won't be able to get rid of Shorthand API support, due to in this case there won't be clear mechanisms for handling `Chat` component's accessibility concerns.

kuzhelov commented on src/themes/teams/components/Chat/chatMessageStyles.ts:
```diff
@@ -18,6 +18,9 @@ const chatMessageStyles: IComponentPartStylesInput<IChatMessageProps, IChatMessa
           marginRight: 'auto',
         }),
     maxWidth: v.messageWidth,
+    width: 'max-content',
```
could you, please, suggest what would be the styling problem if this CSS property won't be applied? The reason I am asking is because this one is marked as experimental: https://developer.mozilla.org/en-US/docs/Web/CSS/width

mnajdova replied:
Without that prop this is how the chat looks:
![image](https://user-images.githubusercontent.com/4512430/46289208-aa688200-c588-11e8-8198-668ac6613440.png)

I will try to implement this behavior using some other styling then...

kuzhelov commented on src/components/Chat/Chat.tsx:
```diff
@@ -35,8 +36,8 @@ class Chat extends UIComponent<Extendable<IChatProps>, any> {
-    /** Shorthand array of messages. */
-    messages: PropTypes.arrayOf(PropTypes.any),
+    /** Shorthand array of the items inside the chat. */
+    items: PropTypes.arrayOf(PropTypes.any),
```
nit: seems that we could be more specific here: 

```js
items:  PropTypes.arrayOf(customPropTypes.itemShorthand)
```

kuzhelov reviewed (commented):
overall the vector of changes taken looks good to me 👍 - few moments that have raised my concerns, left comments for them.

Commit:
-resolved comments on PR (added shorthand example, changed styles for the chat message, provided description of the properties)

kuzhelov commented on docs/src/examples/components/Chat/Types/ChatExample.shorthand.tsx:
```diff
@@ -0,0 +1,51 @@
+      />
+    ),
+  },
+  <Chat.Item>
```
frankly, a bot unsure whether we should provide sort of heterogenous approach in the example - probably, it would be more intuitive if all the items will be the same (especially given that those that are at the end have different content - won't like to mix this differentiation aspect with another one that is about using `Chat.Item` directly instead of content prop)

kuzhelov commented on src/components/Chat/ChatItem.tsx:
```diff
@@ -48,17 +48,11 @@ class ChatItem extends UIComponent<Extendable<IChatItemProps>, any> {
 
     return (
-      <ElementType {...rest} className={cx(classes.root, classes.content)}>
+      <ElementType {...rest} className={classes.root}>
```
👍

Commit:
-providede the content API for the ChatItems in the shorthand example for the Chat

Commit:
Merge branch 'master' into feat/chat-message-types

# Conflicts:
#	CHANGELOG.md

Commit:
-merge conflicts and updating changelog
